src/python/pants/backend/java/dependency_inference/package_parser.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import re
+from dataclasses import dataclass
+
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, DigestContents
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import Sources
+
+
+class ParsedJavaPackage(str):
+    """The declared package for a Java source file.
+
+    An empty string indicates no package declaration, i.e. the unnamed package.  This is generally
+    heavily discouraged or forbidden in non-trivial Java code.
+    """
+
+
+@dataclass(frozen=True)
+class ParseJavaPackageRequest:
+    sources: Sources
+
+
+def regex_parse_java_package(source: str) -> str:
+    """Uses a simple regex to parse the package declaration out of Java source.
+
+    See JLS specification[0] for more details on what valid Java package declaration
+    can look like.
+
+    [0] https://docs.oracle.com/javase/specs/jls/se12/html/jls-7.html#jls-7.4
+    """
+
+    match = re.search(r"package\s+(?P<package_name>[^\s;]+)\s*;", source)
+    if match:
+        return match.group("package_name")
+    return ""
+
+
+@rule
+async def parse_java_package(request: ParseJavaPackageRequest) -> ParsedJavaPackage:
```

Comment:
Is that a restriction we're willing to add?  In general, Java packages don't need to match source directories; that only matters for classfiles on the classpath (or in a jar).  See the various unit tests in the Java subsystem that exercise this for examples

Reply:
Behind an option? Yea, I think so. Well behaved monorepos are very likely to follow it. But because it wouldn't be the default, it's probably not a priority.